### PR TITLE
Meta/markdown comments

### LINKED ISSUE #1420: Table comments visualization

### PR DESCRIPTION
**Pull request by sorja**
Resolves #1420 
Part of #1421 (FE)

image
image

Todo
- [x] Disable XSS attacks (ex. iFrame in markdown) 
- [x] Move fetched data to store (pages.assessmentSection.description)
- [x] Remove references to CKEDITOR #1435 
- [x] Markdown tool: Allow only certain options (remove images? etc)
- [x] Resolve issue where image are not loading in comments (IND, extent of forest, 1a)
image

### CODE REVIEW

File: client/pages/AssessmentSection/Descriptions/CommentableDescription/CommentableDescription.tsx [hunk 1]
```diff
@@ -40,9 +41,7 @@ const CommentableDescription: React.FC<Props> = (props) => {
         />
       </div>
       <div className="fra-description__review-indicator-wrapper">
-        {!disabled && (
-          <pre>{`<ReviewIndicator section={section} title={title} target={[name]} countryIso={countryIso} />`}</pre>
-        )}
+        {!disabled && <ReviewIndicator title={title} topicKey={name} />}
```
Comment (sorja) [inline client/pages/AssessmentSection/Descriptions/CommentableDescription/CommentableDescription.tsx:44]: What should we use here as topicKey?
countryIso + section + name?

Should we add a helper (eg. `Topics.getKey...()`)?

Comment (minotogna) [reply client/pages/AssessmentSection/Descriptions/CommentableDescription/CommentableDescription.tsx:44]: CountyIso + assessmentName + cycleName + sectionName + name 

And yes add 
getKeyComments

Comment (sorja) [reply client/pages/AssessmentSection/Descriptions/CommentableDescription/CommentableDescription.tsx:44]: Under descriptions or topics ?

**Comment by sorja** [comment]
resolved yarn.lock issues with:
git pull origin master-meta
rm -rf yarn.lock
yarn

File: client/pages/AssessmentSection/Descriptions/Description/Description.tsx [hunk 2]
```diff
@@ -21,27 +24,30 @@ const Description: React.FC<Props> = (props) => {
...
   const [open, setOpen] = useState(false)
 
-  const error = user && showAlertEmptyContent && /* !value && */ !loading
-  // let __html = value || template
-  const __html = value
+  const onChange = console.log
+
+  const error = user && showAlertEmptyContent && !value
+  const markdown = value || template
   // if (printView) __html = __html?.split('<p>&nbsp;</p>').join('') // Hack to replace empty lines in print view
```
Comment (minotogna) [inline client/pages/AssessmentSection/Descriptions/Description/Description.tsx:34]: this must be addressed in #1433

File: client/pages/AssessmentSection/Descriptions/Description/useDescription.ts [hunk 3]
```diff
@@ -1,51 +1,14 @@
-type DescriptionState = {
-  loading: boolean
-  onChange: (value: string) => void
-  value: string
+import { useAppSelector } from '@client/store'
+import { useCountryIso } from '@client/hooks'
+
+export default (name: string, section: string, template: string): string => {
```
Comment (minotogna) [inline client/pages/AssessmentSection/Descriptions/Description/useDescription.ts:4]: this hook belong to store

File: client/pages/AssessmentSection/Descriptions/Descriptions.tsx [hunk 4]
```diff
@@ -14,19 +19,18 @@ type Props = {
...
-  //   Objects.isFunction(analysisAndProcessing) ? analysisAndProcessing(state) : analysisAndProcessing,
-  // ]
-  // })
+  const [useNationalData, useAnalysisAndProcessing] = useAppSelector((state) => {
+    if (printOnlyTablesView) {
+      return [false, false]
+    }
+    return [
+      Objects.isFunction(nationalData) ? nationalData(state) : nationalData,
+      Objects.isFunction(analysisAndProcessing) ? analysisAndProcessing(state) : analysisAndProcessing,
```
Comment (minotogna) [inline client/pages/AssessmentSection/Descriptions/Descriptions.tsx:31]: this condition must be verified. I will create a separate issue

File: client/pages/AssessmentSection/Descriptions/Descriptions.tsx [hunk 5]
```diff
@@ -14,19 +19,18 @@ type Props = {
 
 const Descriptions: React.FC<Props> = (props: Props) => {
   const { sectionName, descriptions, disabled } = props
-  const { introductoryText /* nationalData, */ /* analysisAndProcessing */ } = descriptions
-
+  const { introductoryText, nationalData, analysisAndProcessing } = descriptions
   const i18n = useTranslation()
   const [printView, printOnlyTablesView] = [false, false] // TODO: usePrintView()
```
Comment (minotogna) [inline client/pages/AssessmentSection/Descriptions/Descriptions.tsx:24]: this must be addressed in #1433

File: client/pages/OriginalDataPoint/components/Comments/CommentsEditor.tsx [hunk 6]
```diff
@@ -18,6 +17,7 @@ const CommentsEditor: React.FC<Props> = (props) => {
   const textareaRef = useRef<HTMLTextAreaElement>(null)
   const descriptionEditor = useRef(null)
 
+  // TODO: CKEditor removed from project, use MarkdownEditor or MarkdownPreview component
```
Comment (minotogna) [inline client/pages/OriginalDataPoint/components/Comments/CommentsEditor.tsx:20]: we need to use the editor in ODP Comments. separate issue ?

File: client/store/pages/assessmentSection/slice.ts [hunk 7]
```diff
@@ -17,6 +19,7 @@ const initialState: AssessmentSectionState = {
   originalDataPointData: null,
   showOriginalDataPoint: true,
   nodeValueValidation: {},
+  descriptions: [],
```
Comment (minotogna) [inline client/store/pages/assessmentSection/slice.ts:22]: they must be stored by section / name

File: meta/messageCenter/topics.ts [hunk 8]
```diff
@@ -12,8 +12,17 @@ const getMessageBoardChatKey = (countryIso: CountryIso, userA: User, userB: User
   return `${countryKey}_chat_${userKeys}`
 }
 
+const getCommentableDescriptionKey = (
+  countryIso: CountryIso,
+  assessment: Assessment,
+  cycle: Cycle,
+  sectionName: string,
+  name: string
+): string => [countryIso, assessment.props.name, cycle.name, sectionName, name].join('_')
```
Comment (minotogna) [inline meta/messageCenter/topics.ts:21]: prepend `commentable-description`

**Review by minotogna — CHANGES_REQUESTED** [review]

File: client/pages/OriginalDataPoint/components/Comments/CommentsEditor.tsx (hunk 6 above)
Comment (sorja) [reply client/pages/OriginalDataPoint/components/Comments/CommentsEditor.tsx:20]: Separate issue

Comment (minotogna) [reply client/pages/OriginalDataPoint/components/Comments/CommentsEditor.tsx:20]: created #1437

File: client/store/pages/assessmentSection/actions/getDescriptions.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  Record<string, Record<string, string>>,
+  { countryIso: CountryIso; assessmentName: string; cycleName: string; section: string }
+>('section/get/descriptions', async ({ countryIso, assessmentName, cycleName, section }) => {
+  const { data } = await axios.get(ApiEndPoint.Assessment.Data.descriptions(), {
+    params: { countryIso, assessmentName, cycleName, sectionName: section },
+  })
+
+  const res: Record<string, Record<string, string>> = {}
+
+  data.forEach((description: CommentableDescription) => {
```
Comment (minotogna) [inline client/store/pages/assessmentSection/actions/getDescriptions.ts:18]: why not to return the data in that format from the query itself? (just jsonb_build_object)

File: client/store/pages/assessmentSection/hooks/useDescription.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+import { useAppSelector } from '@client/store'
+
+export default (name: string, section: string, template: string): string => {
```
Comment (minotogna) [inline client/store/pages/assessmentSection/hooks/useDescription.ts:3]: pass the args as props:{...} and preferably use sectionName

File: client/store/pages/assessmentSection/slice.ts [hunk 11]
```diff
@@ -67,11 +70,16 @@ export const assessmentSectionSlice = createSlice({
         state.data = TableDatas.updateDatum({ colName, countryIso, tableName, data: state.data, variableName, value })
       })
     })
+
+    builder.addCase(getDescriptions.fulfilled, (state, { payload }) => {
+      state.descriptions = payload
```
Comment (minotogna) [inline client/store/pages/assessmentSection/slice.ts:75]: aren't these grouped by section? let's not override the current data

File: client/store/pages/assessmentSection/stateType.ts [hunk 12]
```diff
@@ -8,4 +8,5 @@ export type AssessmentSectionState = {
   showOriginalDataPoint?: boolean
   // tableName -> nodeUpdate
   nodeValueValidation: Record<string, NodeUpdate>
+  descriptions: Record<string, Record<string, string>>
```
Comment (minotogna) [inline client/store/pages/assessmentSection/stateType.ts:11]: put a little comment like above

File: client/store/pages/assessmentSection/actions/getDescriptions.ts (hunk 9 above)
Comment (sorja) [reply client/store/pages/assessmentSection/actions/getDescriptions.ts:18]: Resolved as: returning single content value in query and handling population in slice

**Review by minotogna — APPROVED** [review]

**Pull request merged**